lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,296 @@
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// an analysis class which relies on manual input locations to mark ops to be
+// hoisted, useful for testing
+class TTIRHoistAnalyzeManual {
+public:
+  using HoistOpSet = std::vector<std::set<mlir::Operation *>>;
+
+  TTIRHoistAnalyzeManual(mlir::Operation *op,
+                         const std::vector<std::string> &targetLocs) {
+    auto moduleOp = llvm::dyn_cast<mlir::ModuleOp>(op);
+    assert(moduleOp && "somehow got non-ModuleOp in TTIRHoistAnalyzeManual!");
+
+    moduleOp.walk([&](mlir::Operation *nestedOp) {
+      if (matchesLocation(nestedOp, targetLocs)) {
+        std::set<mlir::Operation *> opSet;
+        opSet.insert(nestedOp);
+        hoistedOps.push_back(opSet);
+      }
+    });
+  }
+
+  HoistOpSet getResults() { return hoistedOps; }
+
+private:
+  HoistOpSet hoistedOps;
+
+  static bool matchesLocation(mlir::Operation *op,
+                              const std::vector<std::string> &targetLocations) {
+    if (const auto nameLoc = dyn_cast<NameLoc>(op->getLoc())) {
+      const auto &opLocString = nameLoc.getName().str();
+
+      for (const auto &loc : targetLocations) {
+        if (opLocString == loc) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+};
+
+struct TargetLocsParser : llvm::cl::parser<std::vector<std::string>> {
+  TargetLocsParser(llvm::cl::Option &opt)
+      : llvm::cl::parser<std::vector<std::string>>(opt) {}
+
+  bool parse(llvm::cl::Option &opt, llvm::StringRef argName,
+             llvm::StringRef argValue, std::vector<std::string> &value) {
+    // Split the input string on commas
+    llvm::SmallVector<llvm::StringRef, 4> names;
+    argValue.split(names, ',');
+
+    value.assign(names.begin(), names.end());
+
+    return false; // Success
+  }
+
+  static std::string toString(const std::vector<std::string> &locList) {
+    std::string result;
+    for (const auto &loc : locList) {
+      result += loc + ",";
+    }
+    return result;
+  }
+
+  static void print(llvm::raw_ostream &os,
+                    const std::vector<std::string> &locList) {
+    os << "hoist-locs=" << toString(locList) << "\n";
+  }
+};
+
+class TTIRHoistTransform
+    : public impl::TTIRHoistTransformBase<TTIRHoistTransform> {
+public:
+  using impl::TTIRHoistTransformBase<
+      TTIRHoistTransform>::TTIRHoistTransformBase;
+
+  // MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TTIRHoistTransform)
+
+  void runOnOperation() final {
+    mlir::ModuleOp moduleOp = llvm::dyn_cast<mlir::ModuleOp>(getOperation());
+    assert(moduleOp != nullptr && "TTIRHoistTransform should run on ModuleOps, "
```

Comment:
Good point.  To be honest, I got pretty confused by some errors when I used `->` operator instead of `.` operator saying something like `Operation* has no such function X` and wrongly assumed `getOperation()` was returning `Operation*`; I will clean that all up now